Invalidate cache for neo deployments (#2209)
SAP/jenkins-library: documentation/docs/steps/neoDeploy.md

      account: <myDeployAccount>
      host: hana.example.org
```

## Example for invalidating the cache

> Setting parameter `invalidateCache` to `true`, will clean up the cache of the deployed application and reflects the new changes.

Reviewer: This is only true for the an application consumed with a Launchpad site.

> Setting this parameter to true, needs additional configurations to go through. Firstly, create an oauth credential as shown below:

Reviewer: I believe we can help the customer by guiding her through each single step of the way, i.e.:
1. SAP Cloud Platform
2. Left hand side panel choose Security/OAuth
3. Choose Clients tab in OAuth Settings
4. Click Register New Client

If the image 'oauthClientCreation.png' could also show the left hand panel with Security/OAuth settings would be great too, but not a must.

Select the oauth settings from your subaccount, create a new client with subscription to portal/nwc as shown in the following images:

![Oauth client creation](../images/oauthClientCreation.png)

Select the client tab, which provides an option to register a new client as shown below:

> ![Register new client](../images/registerNewOauthClient.png)

Reviewer: This information is basically included in the above image, no need to double it IMO.

> During registering the client, in the subscription tab, select the portal landscape you would like to subscribe to, ex: `portal/nwc` as shown below:

Reviewer: 'portal/nwc' is not shown in the image!

> ![Portal subscription](../images/portalSubscription.png)

Reviewer: This image shows part of the left hand panel, which should be shown earlier. Here is no need for that IMO.

Author: But, I do not see any issue in showing it here as well, Moreover, it gives clear picture IMO.

Reviewer: I believe it adds unnecessary information. Also part of the icons for Repositories, Integration Tokens and Usage Analytics are visible. It is not a cleanly cut image, which does not look professional.

> After saving these changes, a client secret is generated. Ensure to make note of the newly generated secret as it is not available later.

Reviewer: Generating a client secret is wrong here. The user needs to change Authorization Grant to 'Client Credentials' and type a password themselves, only the ID is generated. An image showing this would be helpful.

Author: Good point 👍

> Using the client Id created and client password generated, create a UsernamePassword type credential with client Id as username and client secret as password.

Reviewer: Client ID is generated, password is chosen by the user.

      account: <myDeployAccount>
      host: hana.example.org
```

## Example for invalidating the cache

> * Setting parameter `invalidateCache` to `true`, will clean up the cache of a Fiori Launchpad site, refreshing the content of any applications deployed there. This is **applicable** only when the application is deployed and accessed via Portal service (Fiori Launchpad site).

Reviewer: Why is this a bullet point?

Setting this parameter to true, needs additional configurations to go through. Firstly, create an oauth credential as shown below:

### OAuth credential creation

>   1 Select the oauth settings from your subaccount, create a new client with subscription to portal/nwc as shown in the following images:

Reviewer: Need to add a dot so the md parser understands it's a numbered list. Indentation is then not necessary.
```suggestion
1. Select the oauth settings from your subaccount, create a new client with subscription to portal/nwc as shown in the following images:
```

Reviewer: Use OAuth as this is the official capitalisation

>     ![OAuth client creation](../images/oauthClientCreation.png)

Reviewer: It doesn't seem to find this image. [Check the resulting page here](https://github.com/SAP/jenkins-library/blob/12f450a55ee2131d9468aa6c371d02326704b6c3/documentation/docs/steps/neoDeploy.md#example-for-invalidating-the-cache).

>   2 Select the "Clients" tab, which provides an option to register a new client. Click on "Register New Client" button.

Reviewer: How about: ... Then click on "Register New Client".

>   3 After clicking, in the subscription field, select the portal landscape you would like to subscribe to, ex: `portal/nwc` or `portal/sandbox` as shown below:

Reviewer: Remove "After clicking"

    ![Portal subscription](../images/portalSubscription.png)

  4 In the "Authorization Grant" field, select "Client Credentials" from the drop down menu. Then, enter the "Secret" field with a user defined password and finally save the changes.

> After saving these changes, create a UsernamePassword type credential with client Id (obtained from registering new client) as username and client secret (Password entered in the secret field above) as password in Jenkins.

Reviewer: un-capitalize password

### Set Site Id

After login to the portal service from SAP CP, the user has a possibility to create a site directory, within which the deployed application resides.

> Click on "+" tile, to create a new site directory. After saving the changes to a new site directory tile, the control is navigated to a new page which holds the "site id" in the url. By default, site is set as default as shown below image:

Reviewer: Already commented on how we should not explain how to create a site. We should only explain the default site thing, in my opinion.

![Site set to default](../images/portalSiteSetToDefault.png)

If it is already set to default, no need to configure `siteId` in the configuration file. If not set to default, please configure `siteId`, as shown in the below configuration:

```groovy

> neoDeploy script: this, neo: [credentialsId: 'my-credentials-id', host: hana.example.org, account: 'testAccount']

Reviewer: How about only having the step configured with parameters and not show a config.yml example. I find it confusing with some of the configuration in the step and other configuration in the yaml. How about showing the step configured only with parameters.

Else please add the words: "Example configuration:" as is done in the above example.

Author: @o-liver Is it ok if I improve the Docu in another PR? This is getting so huge that I cannot follow on your comments. If the request is on docu, I would like to handle it in another PR.

      account: <myDeployAccount>
      host: hana.example.org
```

## Example for invalidating the cache

> * Setting parameter `invalidateCache` to `true`, will clean up the cache of the deployed application to portal site and reflects the new changes. This is **applicable** only when the application is deployed and accessed via Portal service (Fiori Launchpad site).

Reviewer: ```suggestion
Setting parameter `invalidateCache` to `true`, will clean up the cache of a Fiori Launchpad site, refreshing the content of any applications deployed there. This is **applicable** only when the application is deployed and accessed via Portal service (Fiori Launchpad site).
```

Setting this parameter to true, needs additional configurations to go through. Firstly, create an oauth credential as shown below:

> * **Oauth credential creation**:

Reviewer: Use OAuth as is the common letter case

> * **Oauth credential creation**:

Reviewer: How about making this a sub-title instead of a bold point with sub-points?

Author: I do not have any issues, but I am not sure, what exact document guidelines we need to follow for this. I would rather prefer bullets, as this is part of the sub heading already.

Reviewer: To each their own. If you like it like this, keep it like this. My idea was going sub-sub by adding another `#` to the heading. I believe in this case three `#` would do the trick.

>   * Select the oauth settings from your subaccount, create a new client with subscription to portal/nwc as shown in the following images:

Reviewer: How about a numbered list, as these steps need to be followed sequentially.

>     ![Oauth client creation](../images/oauthClientCreation.png)

Reviewer: I suggest clicking on the 'Clients' tab already in this image. I like the red frame. OAuth could also be framed red as well as 'Register New Client'

>     ![Oauth client creation](../images/oauthClientCreation.png)

Reviewer: Either show the entire icons on the left hand side or hide them completely. I would prefer showing them, as they guide the user to the right button to click

  * Select the client tab, which provides an option to register a new client as shown below:

>     ![Register new client](../images/registerNewOauthClient.png)

Reviewer: If 'Clients' were already chosen in the above image (oauthClientCreation.png) this image would be redundant.

  * Click on the "Register New Client" button. After clicking, in the subscription field, select the portal landscape you would like to subscribe to, ex:  `portal/nwc` or `portal/sandbox` as shown below:

    ![Portal subscription](../images/portalSubscription.png)

  * In the "Authorization Grant" field, select "Client Credentials" from the drop down menu. Then, enter the "Secret" field with a user defined password and finally save the changes.

> After saving these changes, create an UsernamePassword type credential with client Id (obtained from registering new client) as username and client secret (Password enterred in the Secret field above) as password in Jenkins.

Reviewer: 👍 
typo:
```suggestion
After saving these changes, create a UsernamePassword type credential with client Id (obtained from registering new client) as username and client secret (Password enterred in the Secret field above) as password in Jenkins.
```

> * **Set Site Id**:

Reviewer: Subtitle instead of bold point

> After log in into the portal service from SAP CP, the user has a possibility to create a site directory, within which the deployed application resides. It is created in the following way:

Reviewer: ```suggestion
After login to the portal service from SAP Cloud Platform, the user has a possibility to create a site directory, within which the deployed application resides. It is created in the following way:
```

> ![Site directory](../images/siteDirectory.png)

Reviewer: We do not need to explain how to create a site. If we go down that path we will have to explain too many details. The assumption, as I see it, is that the user already has a Launchpad site with at least one application deployed. We just need to guide them here in case they have not set a default and we need the siteId.
